src/main/java/com/iridium/iridiumskyblock/managers/CooldownProvider.java

```diff
@@ -0,0 +1,92 @@
+package com.iridium.iridiumskyblock.managers;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.lang.NotImplementedException;
+
+/**
+ * Provides various cooldown features.
+ *
+ * @param <T> The type that can have this cooldown
+ */
+public class CooldownProvider<T> {
+
+    private final Map<T, Duration> cooldownTimes = new HashMap<>();
+    private final String name; // Required for database support
+    private final Duration duration;
+    private final boolean persistent; // Required for database support
+
+    /**
+     * The default constructor.
+     *
+     * @param name          The name of this CooldownProvider in the database
+     * @param duration      The duration which cooldowns from this provider have.
+     * @param persistent    True if the cooldowns of this provider should be saved to the database
+     */
+    private CooldownProvider(String name, Duration duration, boolean persistent) {
+        this.name = null;
+        this.duration = duration;
+        this.persistent = persistent;
+    }
+
+    /**
+     * Returns if the specified entity has a valid cooldown.
+     *
+     * @param t The entity which should be checked
+     * @return  True if the entity has a valid cooldown
+     */
+    public boolean isOnCooldown(T t) {
+        return cooldownTimes.containsKey(t) && cooldownTimes.get(t).toMillis() > System.currentTimeMillis();
+    }
+
+    /**
+     * Returns the remaining duration of the cooldown for the provided entity.
+     * ZERO if there is no cooldown.
+     *
+     * @param t The entity which should be checked
+     * @return  The duration of the cooldown, can be ZERO
+     */
+    public Duration getRemainingTime(T t) {
+        if (!isOnCooldown(t)) return Duration.ZERO;
+
+        return cooldownTimes.get(t);
+    }
+
+    /**
+     * Reset the cooldown of the specified entity with the duration of this CooldownProvider.
+     * {@link CooldownProvider#isOnCooldown(Object)} will return true after this.
+     *
+     * @param t The entity which should be checked
+     */
+    public void resetCooldown(T t) {
```

Comment:
not sure I like this naming, reset cooldown gives the impression that you are resetting the cooldown for the player and therefore they will no longer be on cooldown

Reply:
What would you suggest?

Reply:
idk im terrible at naming stuff, just call it `public void i()`

Reply:
apply cooldown? feel like that makes way more sence

Reply:
Sounds worse

Reply:
How so?

Reply:
In my understanding cooldowns are not something that can be applied

Reply:
addCooldown?

Reply:
That was about what I initially planned but the cooldown is not always added because it also resets

Reply:
I pick resetCooldown

Reply:
I'll think about it